components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -284,53 +295,92 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
-                if (!ArrayUtils.isEmpty(matchedUserList)) {
-                    // In the first iteration resultedUserList will be empty.
-                    if (ArrayUtils.isNotEmpty(resultedUserList)) {
-                        // Remove the users from already matched list who are not in the recently matched list.
-                        resultedUserList = getCommonUserEntries(resultedUserList, matchedUserList, key, value);
-                        if (ArrayUtils.isEmpty(resultedUserList)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("No user matched for given claims");
-                            }
-                            return StringUtils.EMPTY;
-                        }
-                    } else {
-                        resultedUserList = matchedUserList;
-                    }
-                } else {
+                if (resultedUserList.size() > 1) {
                     if (log.isDebugEnabled()) {
-                        String message = String.format("No users matched for claim : %s for value : %s", key, value);
-                        log.debug(message);
+                        log.debug("Multiple users matched for given claims set : " +
+                                Arrays.toString(resultedUserList.toArray()));
                     }
-                    return StringUtils.EMPTY;
+                    throw Utils.handleClientException(
+                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_MULTIPLE_MATCHING_USERS, null);
                 }
             }
-        }
-        // Return matched user.
-        if (ArrayUtils.isNotEmpty(resultedUserList) && resultedUserList.length == 1) {
-            return resultedUserList[0];
-        } else {
+            // Return matched user.
+            if (!resultedUserList.isEmpty() && resultedUserList.size() == 1) {
+                return resultedUserList.get(0).getDomainQualifiedUsername();
+            } else {
+                // Return empty when no users are found.
+                return StringUtils.EMPTY;
+            }
+        } catch (org.wso2.carbon.user.core.UserStoreException e) {
             if (log.isDebugEnabled()) {
-                log.debug("Multiple users matched for given claims set : " + Arrays.toString(resultedUserList));
+                log.debug("Error while getting users from user store for the given claim set: " +
+                        Arrays.toString(claims.keySet().toArray()));
             }
-            throw Utils
-                    .handleClientException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_MULTIPLE_MATCHING_USERS,
-                            null);
+            throw new IdentityRecoveryException(e.getErrorCode(), e.getMessage());
+        } catch (UserStoreException e) {
+            throw new IdentityRecoveryException(e.getMessage());
```

Comment:
Throw the throwable also. otherwise the exception stacktrace will be lost.

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/ec87914bdfabae3c2012cda51e59249fadeefe94